File: Android/iOrder/app/src/main/java/hr/fer/oobl/iorder/iorder/injection/application/ApplicationComponent.java
package hr.fer.oobl.iorder.iorder.injection.application;

import javax.inject.Singleton;

import dagger.Component;
import hr.fer.oobl.iorder.iorder.base.IOrderApplication;
import hr.fer.oobl.iorder.iorder.injection.application.modules.ApplicationModule;
import hr.fer.oobl.iorder.iorder.injection.application.modules.DataModule;
import hr.fer.oobl.iorder.iorder.injection.application.modules.DomainModule;
import hr.fer.oobl.iorder.iorder.injection.application.modules.MappersModule;
import hr.fer.oobl.iorder.iorder.injection.application.modules.ThreadingModule;
import hr.fer.oobl.iorder.iorder.injection.application.modules.UseCaseModule;

@Singleton
@Component(
        modules = {
                ApplicationModule.class,
                DataModule.class,
                DomainModule.class,
                MappersModule.class,
                ThreadingModule.class,
                UseCaseModule.class,
        }
)
public interface ApplicationComponent extends ApplicationComponentInjects, ApplicationComponentExposes {

    final class Initializer {

        static public ApplicationComponent init(final IOrderApplication iOrderApplication) {
            return DaggerApplicationComponent.builder()
                                             .applicationModule(new ApplicationModule(iOrderApplication))
                                             .dataModule(new DataModule(iOrderApplication))
                                             .mappersModule(new MappersModule())
                                             .domainModule(new DomainModule())
                                             .threadingModule(new ThreadingModule())
                                             .useCaseModule(new UseCaseModule())
                                             .build();
        }

        private Initializer() {
        }
    }
}

File: Android/iOrder/app/src/main/java/hr/fer/oobl/iorder/iorder/injection/application/ApplicationComponent.kt
package hr.fer.oobl.iorder.iorder.injection.application

import javax.inject.Singleton

import dagger.Component
import hr.fer.oobl.iorder.iorder.application.IOrderApplication
import hr.fer.oobl.iorder.iorder.injection.application.modules.ApplicationModule
import hr.fer.oobl.iorder.iorder.injection.application.modules.DataModule
import hr.fer.oobl.iorder.iorder.injection.application.modules.DomainModule
import hr.fer.oobl.iorder.iorder.injection.application.modules.MappersModule
import hr.fer.oobl.iorder.iorder.injection.application.modules.ThreadingModule
import hr.fer.oobl.iorder.iorder.injection.application.modules.UseCaseModule

@Singleton
@Component(modules = arrayOf(ApplicationModule::class, DataModule::class, DomainModule::class, MappersModule::class, ThreadingModule::class, UseCaseModule::class))
interface ApplicationComponent : ApplicationComponentInjects, ApplicationComponentExposes {

    object Initializer {

        fun init(iOrderApplication: IOrderApplication): ApplicationComponent {
            return DaggerApplicationComponent.builder()
                    .applicationModule(ApplicationModule(iOrderApplication))
                    .dataModule(DataModule(iOrderApplication))
                    .mappersModule(MappersModule())
                    .domainModule(DomainModule())
                    .threadingModule(ThreadingModule())
                    .useCaseModule(UseCaseModule())
                    .build()
        }
    }
}
